File: backend/utils/process_my_excel.py
"""
Direct Excel File Processor
Process your existing Excel files from student_list_excel folder
"""

import os
from pathlib import Path
from main import StudentDatabase, StudentDataExtractor

def main():
    print("="*70)
    print("🎓 STUDENT LIST EXCEL PROCESSOR")
    print("="*70)
    
    # Your actual path
    excel_folder = r"C:\Users\vince pascua\Downloads\vscodes\mongo\MongoDB\backend\utils\uploaded_files\student_list_excel"
    
    print(f"\n📁 Excel folder: {excel_folder}")
    
    # Check if folder exists
    if not os.path.exists(excel_folder):
        print(f"❌ Folder not found!")
        print(f"ℹ️  Please check the path")
        return
    
    # Find Excel files
    excel_files = list(Path(excel_folder).glob("*.xlsx")) + \
                 list(Path(excel_folder).glob("*.xls"))
    
    if not excel_files:
        print(f"\n⚠️  No Excel files found in this folder")
        print(f"\nℹ️  Files in folder:")
        try:
            all_files = list(Path(excel_folder).iterdir())
            for file in all_files:
                print(f"   - {file.name}")
        except Exception as e:
            print(f"   Error: {e}")
        return
    
    print(f"\n✅ Found {len(excel_files)} Excel file(s):")
    for i, file in enumerate(excel_files, 1):
        size_kb = file.stat().st_size / 1024
        print(f"   {i}. {file.name} ({size_kb:.1f} KB)")
    
    # Connect to MongoDB
    print(f"\n🔌 Connecting to MongoDB...")
    try:
        db = StudentDatabase()
        print("✅ Connected to MongoDB successfully")
    except Exception as e:
        print(f"❌ MongoDB connection failed: {e}")
        print("\n💡 Make sure MongoDB is running:")
        print("   net start MongoDB")
        return
    
    # Show current stats
    stats = db.get_statistics()
    print(f"\n📊 Current Database:")
    print(f"   Total Students: {stats['total_students']}")
    print(f"   Pending Media: {stats['pending_media']}")
    
    # Ask to clear existing data
    if stats['total_students'] > 0:
        clear = input(f"\n⚠️  Found {stats['total_students']} existing students. Clear them first? (yes/no): ").strip().lower()
        if clear == 'yes':
            db.clear_all_data()
            print("✅ Existing data cleared")
    
    # Process confirmation
    print(f"\n{'='*70}")
    confirm = input(f"🚀 Process all {len(excel_files)} file(s)? (yes/no): ").strip().lower()
    
    if confirm != 'yes':
        print("❌ Operation cancelled")
        db.close()
        return
    
    # Process each file
    print(f"\n{'='*70}")
    print("🔄 PROCESSING FILES...")
    print('='*70)
    
    total_processed = 0
    total_students = 0
    
    for i, excel_file in enumerate(excel_files, 1):
        print(f"\n[{i}/{len(excel_files)}] 📄 Processing: {excel_file.name}")
        print("-"*70)
        
        try:
            # Get student count before
            before_count = db.get_statistics()['total_students']
            
            # Process file
            success = StudentDataExtractor.process_excel(str(excel_file), db)
            
            # Get student count after
            after_count = db.get_statistics()['total_students']
            students_added = after_count - before_count
            
            if success and students_added > 0:
                print(f"✅ Success! Added {students_added} students")
                total_processed += 1
                total_students += students_added
            elif success:
                print(f"⚠️  File processed but no students added")
            else:
                print(f"❌ Failed to process file")
                
        except Exception as e:
            print(f"❌ Error: {e}")
            # Show detailed error for debugging
            import traceback
            print("\nDetailed error:")
            traceback.print_exc()
    
    # Final summary
    print(f"\n{'='*70}")
    print("✅ PROCESSING COMPLETE!")
    print('='*70)
    print(f"Files processed: {total_processed}/{len(excel_files)}")
    print(f"Total students added: {total_students}")
    
    # Show final stats
    final_stats = db.get_statistics()
    print(f"\n📊 Final Database Statistics:")
    print(f"   Total Students: {final_stats['total_students']}")
    print(f"   Pending Media: {final_stats['pending_media']}")
    print(f"   Average Completion: {final_stats['average_completion']:.1f}%")
    
    if final_stats['by_department']:
        print(f"\n📚 Students by Department:")
        for dept, count in final_stats['by_department'].items():
            dept_name = {
                'CCS': 'Computer Studies',
                'CHTM': 'Hospitality & Tourism',
                'CBA': 'Business Administration',
                'CTE': 'Teacher Education',
                'UNKNOWN': 'Unclassified'
            }.get(dept, dept)
            print(f"   • {dept_name}: {count} students")
    
    # Show pending media
    print(f"\n⏳ Students Waiting for Media:")
    pending = db.get_pending_media_students()
    if pending:
        print(f"   {len(pending)} students need image/audio uploads")
        
        show_list = input(f"\n   Show list? (yes/no): ").strip().lower()
        if show_list == 'yes':
            for student in pending[:10]:  # Show first 10
                print(f"   - {student.get('full_name', 'N/A')} ({student['student_id']})")
            if len(pending) > 10:
                print(f"   ... and {len(pending)-10} more")
    else:
        print("   None! (All students have complete media)")
    
    db.close()
    print(f"\n👋 Disconnected from MongoDB")
    print("="*70)

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\n⚠️  Process interrupted by user")
    except Exception as e:
        print(f"\n❌ Unexpected error: {e}")
        import traceback
        traceback.print_exc()
    
    input("\nPress Enter to exit...")

File: express-backend/utils/process_my_excel.js
/**
 * Direct Excel File Processor
 * Process your existing Excel files from student_list_excel folder
 */

const readline = require('readline');
const path = require('path');
const fs = require('fs').promises;
const { StudentDatabase, StudentDataExtractor } = require('./main');

class ExcelProcessor {
  constructor() {
    // Create readline interface for user input
    this.rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });
  }

  // Helper function to prompt user for input
  prompt(question) {
    return new Promise((resolve) => {
      this.rl.question(question, (answer) => {
        resolve(answer);
      });
    });
  }

  async main() {
    console.log('='.repeat(70));
    console.log('🎓 STUDENT LIST EXCEL PROCESSOR');
    console.log('='.repeat(70));

    // Path for files in utils/uploaded_files/student_list_excel
    const excelFolder = path.join(__dirname, 'uploaded_files', 'student_list_excel');
    console.log(`\n📁 Excel folder: ${excelFolder}`); 
    // excelFolder → the directory path where Excel files are stored

    // Check if folder exists
    try {
      await fs.access(excelFolder);
    } catch {
      console.log('❌ Folder not found!');
      console.log('ℹ️  Creating folder...');
      await fs.mkdir(excelFolder, { recursive: true });
      console.log(`✅ Folder created: ${excelFolder}`); 
      // excelFolder → shows newly created folder path
      console.log('ℹ️  Place your Excel files there and run again');
      this.rl.close();
      return;
    }

    // Find Excel files
    let excelFiles = [];
    try {
      const files = await fs.readdir(excelFolder);
      excelFiles = files
        .filter(file => file.endsWith('.xlsx') || file.endsWith('.xls'))
        .map(file => path.join(excelFolder, file));
    } catch (error) {
      console.error(`❌ Error reading folder: ${error.message}`); 
      // error.message → reason why folder reading failed
      this.rl.close();
      return;
    }

    if (excelFiles.length === 0) {
      console.log('\n⚠️  No Excel files found in this folder');
      console.log('\nℹ️  Files in folder:');
      try {
        const allFiles = await fs.readdir(excelFolder);
        allFiles.forEach(file => {
          console.log(`   - ${file}`); 
          // file → lists each filename in folder
        });
      } catch (error) {
        console.log(`   Error: ${error.message}`); 
        // error.message → describes folder reading issue
      }
      this.rl.close();
      return;
    }

    console.log(`\n✅ Found ${excelFiles.length} Excel file(s):`); 
    // excelFiles.length → number of Excel files found

    for (let i = 0; i < excelFiles.length; i++) {
      const filePath = excelFiles[i];
      const fileName = path.basename(filePath);
      try {
        const stats = await fs.stat(filePath);
        const sizeKb = (stats.size / 1024).toFixed(1);
        console.log(`   ${i + 1}. ${fileName} (${sizeKb} KB)`); 
        // i → index of file in the list
        // fileName → name of the Excel file
        // sizeKb → size of the file in kilobytes
      } catch (error) {
        console.log(`   ${i + 1}. ${fileName} (size unknown)`); 
        // fileName → name shown even if file size failed to load
      }
    }

    // Connect to MongoDB
    console.log('\n🔌 Connecting to MongoDB...');
    let db;
    try {
      db = new StudentDatabase();
      await db.connect();
      console.log('✅ Connected to MongoDB successfully');
    } catch (error) {
      console.error(`❌ MongoDB connection failed: ${error.message}`); 
      // error.message → reason for MongoDB connection failure
      console.log('\n💡 Make sure MongoDB is running:');
      console.log('   net start MongoDB');
      this.rl.close();
      return;
    }

    // Show current stats
    let stats = await db.getStatistics();
    console.log('\n📊 Current Database:');
    console.log(`   Total Students: ${stats.total_students}`); 
    // stats.total_students → total count of students in database
    console.log(`   Pending Media: ${stats.pending_media}`); 
    // stats.pending_media → number of students missing media files

    // Ask to clear existing data
    if (stats.total_students > 0) {
      const clear = await this.prompt(
        `\n⚠️  Found ${stats.total_students} existing students. Clear them first? (yes/no): `
      );
      if (clear.trim().toLowerCase() === 'yes') {
        await db.clearAllData();
        console.log('✅ Existing data cleared');
      }
    }

    // Process confirmation
    console.log('\n' + '='.repeat(70));
    const confirm = await this.prompt(
      `🚀 Process all ${excelFiles.length} file(s)? (yes/no): `
    );
    // excelFiles.length → total number of files to be processed

    if (confirm.trim().toLowerCase() !== 'yes') {
      console.log('❌ Operation cancelled');
      await db.close();
      this.rl.close();
      return;
    }

    // Process each file
    console.log('\n' + '='.repeat(70));
    console.log('📄 PROCESSING FILES...');
    console.log('='.repeat(70));

    let totalProcessed = 0; // total number of files successfully processed
    let totalStudents = 0;  // total number of students added

    for (let i = 0; i < excelFiles.length; i++) {
      const excelFile = excelFiles[i];
      const fileName = path.basename(excelFile);

      console.log(`\n[${i + 1}/${excelFiles.length}] 📄 Processing: ${fileName}`); 
      // i + 1 → current file index
      // excelFiles.length → total files count
      // fileName → name of the current file
      console.log('-'.repeat(70));

      try {
        // Get student count before
        const beforeStats = await db.getStatistics();
        const beforeCount = beforeStats.total_students;

        // Process file
        const success = await StudentDataExtractor.processExcel(excelFile, db);

        // Get student count after
        const afterStats = await db.getStatistics();
        const afterCount = afterStats.total_students;
        const studentsAdded = afterCount - beforeCount;

        if (success && studentsAdded > 0) {
          console.log(`✅ Success! Added ${studentsAdded} students`); 
          // studentsAdded → number of new students from this file
          totalProcessed++;
          totalStudents += studentsAdded;
        } else if (success) {
          console.log('⚠️  File processed but no students added');
        } else {
          console.log('❌ Failed to process file');
        }

      } catch (error) {
        console.error(`❌ Error: ${error.message}`); 
        // error.message → reason for failure
        console.log('\nDetailed error:');
        console.error(error.stack); 
        // error.stack → full error trace
      }
    }

    // Final summary
    console.log('\n' + '='.repeat(70));
    console.log('✅ PROCESSING COMPLETE!');
    console.log('='.repeat(70));
    console.log(`Files processed: ${totalProcessed}/${excelFiles.length}`); 
    // totalProcessed → successfully processed files
    // excelFiles.length → total number of files
    console.log(`Total students added: ${totalStudents}`); 
    // totalStudents → number of new student records added

    // Show final stats
    const finalStats = await db.getStatistics();
    console.log('\n📊 Final Database Statistics:');
    console.log(`   Total Students: ${finalStats.total_students}`); 
    // finalStats.total_students → final total in database
    console.log(`   Pending Media: ${finalStats.pending_media}`); 
    // finalStats.pending_media → number needing media files
    console.log(`   Average Completion: ${finalStats.average_completion.toFixed(1)}%`); 
    // finalStats.average_completion → average data completeness

    if (Object.keys(finalStats.by_department).length > 0) {
      console.log('\n📚 Students by Department:');
      
      const deptNames = {
        'CCS': 'Computer Studies',
        'CHTM': 'Hospitality & Tourism',
        'CBA': 'Business Administration',
        'CTE': 'Teacher Education',
        'UNKNOWN': 'Unclassified'
      };

      Object.entries(finalStats.by_department).forEach(([dept, count]) => {
        const deptName = deptNames[dept] || dept;
        console.log(`   • ${deptName}: ${count} students`); 
        // deptName → department name
        // count → student count per department
      });
    }

    // Show pending media
    console.log('\n⏳ Students Waiting for Media:');
    const pending = await db.getPendingMediaStudents();
    if (pending.length > 0) {
      console.log(`   ${pending.length} students need image/audio uploads`); 
      // pending.length → total number of incomplete students

      const showList = await this.prompt('\n   Show list? (yes/no): ');
      if (showList.trim().toLowerCase() === 'yes') {
        const displayCount = Math.min(pending.length, 10);
        for (let i = 0; i < displayCount; i++) {
          const student = pending[i];
          console.log(`   - ${student.full_name || 'N/A'} (${student.student_id})`); 
          // student.full_name → student's full name
          // student.student_id → student ID number
        }
        if (pending.length > 10) {
          console.log(`   ... and ${pending.length - 10} more`); 
          // pending.length - 10 → number of remaining students not displayed
        }
      }
    } else {
      console.log('   None! (All students have complete media)');
    }

    await db.close();
    console.log('\n👋 Disconnected from MongoDB');
    console.log('='.repeat(70));

    this.rl.close();
  }

  async run() {
    try {
      await this.main();
    } catch (error) {
      console.error(`\n❌ Unexpected error: ${error.message}`); 
      // error.message → unexpected runtime error
      console.error(error.stack); 
      // error.stack → full error details
    } finally {
      await this.prompt('\nPress Enter to exit...');
      this.rl.close();
    }
  }
}

// Handle Ctrl+C gracefully
process.on('SIGINT', () => {
  console.log('\n\n⚠️  Process interrupted by user');
  process.exit(0);
});

// Run if executed directly
if (require.main === module) {
  const processor = new ExcelProcessor();
  processor.run();
}

module.exports = ExcelProcessor;